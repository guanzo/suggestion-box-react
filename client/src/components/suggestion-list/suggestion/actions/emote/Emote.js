import React, { PureComponent } from 'react';
import { connect } from 'react-redux'
import { postEmote } from '@store/emotes'
import Reactions from './Reactions'
import PopupManager from './PopupManager'
import './Emote.scss'


class Emote extends PureComponent {
    render() {
		let { emotes, emoteReactions } = this.props
		let onSelect = { onSelectEmote: this.onSelectEmote }
		let isAllowedToEmote = this.isAllowedToEmote()

		let reactionProps = {
			emoteReactions,
			...onSelect,
			isAllowedToEmote
		}
		let popupProps = {
			togglePopup: this.togglePopup,
			closePopup: this.closePopup,
			emotes,
			...onSelect
		}
		
        return (
            <React.Fragment>
				<Reactions {...reactionProps}></Reactions>
				{ isAllowedToEmote ? <PopupManager {...popupProps}></PopupManager> : null }
			</React.Fragment>
        )
	}
	isAllowedToEmote(){
		let { hasEmoted, currentUser } = this.props
		return !(hasEmoted || currentUser.isAnonymousUser)
	}//can select emote from reactions list or popup
	onSelectEmote = (emoteId)=>{
		let { postEmote, id: suggestionId, listType } = this.props
		if(!this.isAllowedToEmote())
			return;

		postEmote(suggestionId, emoteId, listType)
	}
	componentDidCatch(){
		
	}
}
export { Emote }

const mapStateToProps = (state, ownProps) => {
    return {
		emotes: state.emotes
    }
}
const mapDispatchToProps = (dispatch, ownProps) => {
    return {
		postEmote: (...args)=>dispatch(postEmote(...args)),
    }
}
export default connect(mapStateToProps,mapDispatchToProps)(Emote)
