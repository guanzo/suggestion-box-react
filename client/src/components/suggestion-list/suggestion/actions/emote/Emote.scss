$emote-size: 20px;
$popup-width: 290px;
$popup-height: 150px;

@mixin emote-wrapper {
	border: 1px solid lightgrey;
	border-radius: 5px;
	@include flex-center;
	margin: 2px;
	padding: 0px 1px;
	cursor: pointer;
}

@mixin reaction-clickable {
	transition: .35s;
	&:hover{
		border: 1px solid $primary;
	}
}

.suggestion-actions {
	position: relative;
	.emote-image-wrapper{
		height: $emote-size;
		width: $emote-size;
		margin: 2px;
		@include flex-center;
	}
	.emote-image {
		height: $emote-size;
		width: $emote-size;
		object-fit: contain;
		cursor: pointer;
	}
	.reaction{
		@include emote-wrapper;
		position: relative;
		.reaction-count{
			margin: 2px;
			font-size: 0.75rem;
		}
		&.reaction-clickable{
			@include reaction-clickable;
			&:after {
				@include flex-center;
				content:'+1';
				background: white;
				border-radius: 5px;
				width: 100%;
				height: 100%;
				opacity: 0;
				position: absolute;
				pointer-events: none;
				font-size: 0.9rem;
				transition: 0.35s;
			}
			&:hover:after{
				opacity: 1;
			}
		}
	}
	.open-emote-explorer,
	.show-all-emotes{
		@include emote-wrapper;
		@include reaction-clickable;
		padding: 0px 2px;
	}
	.popper{
		z-index: 10;
		width: $popup-width;
		height: $popup-height;
	}	
}

.emote-explorer{
	display: flex;
	flex-direction: column;
	padding: 10px;
	width: 100%;
	height: 100%;
	background: white;
	box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
	
	.emote-grid{
		display: flex;
		justify-content: center;
		flex-wrap: wrap;
		padding: 5px 0px;
	}
	.field{
		margin-top: auto;
	}
}