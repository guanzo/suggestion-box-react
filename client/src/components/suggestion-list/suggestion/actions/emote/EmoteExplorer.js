import React, { PureComponent } from 'react';
import _ from 'lodash'
import EmoteImage from './EmoteImage'

class EmoteExplorer extends PureComponent {
	constructor(props){
		super(props)
		this.state = {
			query: ''
		}
	}
    render() {
		let { query } = this.state
		let { onSelectEmote } = this.props
		let emotes = this.searchedEmotes()
        return (
            <div className="emote-explorer">
				<div className="subtitle is-size-6 has-text-centered m-b-0">
					Emote reactions
				</div>
				<div className="emote-grid">
				{emotes.map(emote=>(
					<EmoteImage emote={emote}
						onSelectEmote={onSelectEmote} 
						key={emote.id}
					></EmoteImage>
				))}
				</div>
				<div className="field">
					<div className="control is-expanded">
						<input className="input is-primary is-small" 
							onChange={this.handleInput} 
							value={query} 
							placeholder='Search for Emotes'
						/>
					</div>
				</div>
			</div>
        )
	}
	handleInput = (e)=>{
		this.setState({ query: e.target.value.trim() })
	}
	searchedEmotes(){
		const maxEmotes = 33
		let { query } = this.state
		let { emotes } = this.props
		//escape regex special characters
		query = query.replace(/[.?*+^$[\]\\(){}|-]/g, "\\$&");
		let regex = new RegExp(query,'i');
		return _(emotes)
				.filter(d=>regex.test(d.code))
				.take(maxEmotes)
				.value()
	}
}

export default EmoteExplorer
