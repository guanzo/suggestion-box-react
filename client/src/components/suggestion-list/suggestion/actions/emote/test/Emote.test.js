
import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Emote } from '../Emote';
import EmoteExplorer from '../EmoteExplorer';


function mockProps({ hasEmoted = false, isAnonymousUser = false } = {}){
	return {
		hasEmoted,
		currentUser: {
			isAnonymousUser
		},
		emotes:[],
		emoteReactions:[],
	}
}

describe('<Emote />',()=>{
	it('does not show emote explorer to anon users',()=>{
		let props = { isAnonymousUser: true }
		const wrapper = shallow(<Emote {...mockProps(props)} />)
		expect(wrapper.find(EmoteExplorer)).toHaveLength(0);
	})
	it('does not show emote explorer to users who have already emoted',()=>{
		let props = { hasEmoted: true }
		const wrapper = shallow(<Emote {...mockProps(props)} />)
		expect(wrapper.find(EmoteExplorer)).toHaveLength(0);
	})
	it('renders correctly',()=>{
		const component = renderer.create(<Emote {...mockProps()}/>)
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	})
})