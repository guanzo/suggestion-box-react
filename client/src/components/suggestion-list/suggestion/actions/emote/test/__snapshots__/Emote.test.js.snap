// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Emote /> renders correctly 1`] = `
<div
  className="popup-manager"
>
  <div
    className="open-emote-explorer"
    onClick={[Function]}
  >
    <div>
      +
    </div>
    <div
      className="icon is-small"
    >
      <i
        className="fa fa-smile-o"
      />
    </div>
  </div>
</div>
`;
